accounts/abi/bind/bind.go

```diff
@@ -0,0 +1,395 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package bind generates Ethereum contract Go bindings.
+package bind
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"golang.org/x/tools/imports"
+)
+
+// Bind generates a Go wrapper around a contract ABI. This wrapper isn't meant
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention opposed to having to
+// manually maintain hard coded strings that break on runtime.
+func Bind(abijson string, bytecode string, pkg string, kind string) (string, error) {
+	// Parse the actual ABI to generate the binding for
+	abi, err := abi.JSON(strings.NewReader(abijson))
+	if err != nil {
+		return "", err
+	}
+	// Generate the contract type, fields and methods
+	code := new(bytes.Buffer)
+	kind = strings.ToUpper(kind[:1]) + kind[1:]
+	fmt.Fprintf(code, "%s\n", bindContract(kind, strings.TrimSpace(abijson), strings.TrimSpace(bytecode)))
+
+	methods := make([]string, 0, len(abi.Methods))
+	for name, _ := range abi.Methods {
+		methods = append(methods, name)
+	}
+	sort.Strings(methods)
+
+	for _, method := range methods {
+		fmt.Fprintf(code, "%s\n", bindMethod(kind, abi.Methods[method]))
+	}
+	// Format the code with goimports and return
+	buffer := new(bytes.Buffer)
+
+	fmt.Fprintf(buffer, "// This file is an automatically generated Go binding based on the contract ABI\n")
+	fmt.Fprintf(buffer, "// defined in %sABI. Do not modify as any change will likely be lost!\n\n", kind)
+	fmt.Fprintf(buffer, "package %s\n\n", pkg)
+	fmt.Fprintf(buffer, "%s\n\n", string(code.Bytes()))
+
+	blob, err := imports.Process("", buffer.Bytes(), nil)
+	if err != nil {
+		return "", fmt.Errorf("%v\n%s", err, code)
+	}
+	return string(blob), nil
+}
+
+// bindContract generates the basic wrapper code for interacting with an Ethereum
+// contract via the abi package. All contract methods will call into the generic
+// ones generated here.
+func bindContract(kind string, abijson string, bytecode string) string {
+	code := ""
+
+	// Generate the hard coded ABI and byte code used for Ethereum interaction
+	code += fmt.Sprintf("// Ethereum ABI used to generate the binding from.\nconst %sABI = `%s`\n\n", kind, abijson)
```

Comment:
Wouldn't it be much more readable if you used package `text/template` here? This would allow you to do something like:

```go
// Ethereum ABI used to generate the
const {{kind}}ABI = {{abi_json}}

{{if bytecode}}
// Ethereum VM bytecode ..
const {{kind}}ABI = {{bytecode}}
{{end}}

... and so forth ...
```